### PR TITLE
284 withdraw transfer

### PR DESCRIPTION
**Pull request by lapinlapin**
#### What's this PR do?
Created flow of Withdraw
#### Where should the reviewer start?
/views/Withdraw
#### How should this be manually tested?
http://localhost:8080/
#### What are the relevant tickets?
#284 
#### URL
Provide a link to your deploy from http://staging-ui.trusty.apasia.tech:8080/branch-284-withdraw-transfer

### CODE REVIEW

File: src/store/modules/withdraw.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+const getters = {
+  modalDisplayed: state => state.showModal,
+  getWithdrawAsset: state => state.withdrawAsset,
+  getWithdrawAmount: state => state.withdrawAmount,
+  getWithdrawAddress: state => state.withdrawAddress,
+  getWithdrawStep: state => state.withdrawStep
+}
+
+const mutations = {
+  [types.TOGGLE_WITHDRAW_MODAL](state, value) {
```
Comment (roma219) [inline src/store/modules/withdraw.js]: value not used

File: src/views/Withdraw/WithdrawConfirmAddress.vue [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    Button,
+    SimpleInput
+  },
+  data() {
+    return {
+      address: '',
+      validUser: false,
+      userLoaded: false,
+      errorTitle: 'user not found',
+      validateUser: debounce(this.getUser, 500)
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAddress.vue]: Don't store functions in data. Create this function in `created`

File: src/views/Withdraw/WithdrawConfirmAddress.vue [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    ...mapActions({
+      setWithdrawStep: 'withdraw/setWithdrawStep',
+      setWithdrawAddress: 'withdraw/setWithdrawAddress'
+    }),
+    async getUser() {
+      this.validUser = false
+      this.userLoaded = false
+      const user = await getUser(this.address)
+      this.userLoaded = true
+      if (user.success) {
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAddress.vue]: `this.validUser = user.success`

File: src/views/Withdraw/WithdrawConfirmAmount.vue [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    }
+  },
+  computed: {
+    ...mapGetters({
+      withdrawAsset: 'withdraw/getWithdrawAsset'
+    }),
+    maxWithdrawTitle() {
+      return `max ${this.withdrawAsset.tokens} ${this.withdrawAsset.tiker}`
+    },
+    isAmountValid() {
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAmount.vue]: This function should return `true/false` not error string, use another computed if needed

File: src/views/Withdraw/WithdrawConfirmAmount.vue [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      const isNumeric = () => !isNaN(parseFloat(this.amount)) && isFinite(this.amount)
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAmount.vue]: there is a `type="number"` prop in simple input (see how it is done in NewOrder) that will only allow numbers

File: src/views/Withdraw/WithdrawConfirmAmount.vue [hunk 6]
```diff
@@ -0,0 +1,126 @@
+<template>
+  <div class="withdraw-container h-full sm:w-120">
+    <div class="withdraw-title">Withdraw</div>
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAmount.vue]: It is better to be placed in the top component, now you repeat it & it's style in every step. Same goes to container with sm:w-120

File: src/views/Withdraw/WithdrawWidget.vue [hunk 7]
```diff
@@ -0,0 +1,42 @@
+<template>
+  <Modal
```
Comment (roma219) [inline src/views/Withdraw/WithdrawWidget.vue:2]: title and container should be here to prevent a lot of code copy-paste

**Review by roma219 — CHANGES_REQUESTED** [review]
Needs some optimizations

File: src/views/Withdraw/WithdrawConfirmAmount.vue (hunk 6 above)
Comment (lapinlapin) [reply src/views/Withdraw/WithdrawConfirmAmount.vue]: i use scoped, therefore so

**Comment by lapinlapin** [comment]
@roma219 fixed, except the last

File: src/views/Withdraw/Withdraw.vue [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    }),
+    selectAsset(asset) {
+      this.setWithdrawAsset({ asset })
+      this.setWithdrawStep('withdrawConfirmAmount')
+    }
+  }
+}
+</script>
+<style lang="scss">
+  .withdraw-container {
```
Comment (roma219) [inline src/views/Withdraw/Withdraw.vue]: no such class here

File: src/views/Withdraw/Withdraw.vue [hunk 9]
```diff
@@ -0,0 +1,89 @@
+<template>
+  <div class="withdraw-form">
+    <div class="withdraw-sub-title">Choose token</div>
+    <div class="withdraw-loader-wrapper">
+      <LoadingContainer :loading="!coinslist.length">
+        <ScrollingContainer
+          :shadower-height="0"
```
Comment (roma219) [inline src/views/Withdraw/Withdraw.vue]: should have a shadower

File: src/views/Withdraw/Withdraw.vue [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    ScrollingContainer,
+    LoadingContainer,
+    Button
+  },
+  computed: {
+    ...mapGetters({
+      coins: 'portfolio/getItems'
+    }),
+    coinslist() {
+      return this.coins.filter(coin => coin.tokens > 0)
```
Comment (roma219) [inline src/views/Withdraw/Withdraw.vue]: `coin => coin.tokens`

File: src/views/Withdraw/WithdrawConfirmAmount.vue [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+      return `max ${this.withdrawAsset.tokens} ${this.withdrawAsset.tiker}`
+    },
+    isAmountValid() {
+      if (+this.amount < 1 || +this.amount > this.withdrawAsset.tokens) {
+        return true
+      }
+      return false
+    },
+    error() {
+      if (this.amount === undefined || this.amount === '') {
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAmount.vue:68]: if (!this.amount) return

**Review by roma219 — CHANGES_REQUESTED** [review]
1)
image
2) 
image
3) Should be enabled on mobile as well
image
4) Confirm should be disabled when checking user is in process
image

File: src/views/Withdraw/WithdrawConfirmAmount.vue (hunk 11 above)
Comment (lapinlapin) [reply src/views/Withdraw/WithdrawConfirmAmount.vue:68]: no)

Comment (roma219) [reply src/views/Withdraw/WithdrawConfirmAmount.vue:68]: well not return, but still

File: src/views/Withdraw/WithdrawConfirmAddress.vue [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAddress.vue]: No need for this condition, straight `this.validUser = user.success` as it's just gonna make it true/false

File: src/views/Withdraw/WithdrawConfirmAmount.vue [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+  },
+  computed: {
+    ...mapGetters({
+      withdrawAsset: 'withdraw/getWithdrawAsset'
+    }),
+    maxWithdrawTitle() {
+      return `max ${this.withdrawAsset.tokens} ${this.withdrawAsset.tiker}`
+    },
+    isAmountValid() {
+      if (+this.amount <= 0 || +this.amount > this.withdrawAsset.tokens) {
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAmount.vue]: Can make it even shorter:
`return !!(+this.amount <= 0 || +this.amount > this.withdrawAsset.tokens)`

File: src/views/Withdraw/WithdrawFinish.vue [hunk 14]
```diff
@@ -0,0 +1,64 @@
+<template>
+  <div class="withdraw-form">
+    <div class="withdraw-content">Are you sure you want to withdraw {{ withdrawAmount }} {{ withdrawAsset.tiker.toUpperCase() }} to {{ withdrawAddress }} address?</div>
+    <div class="withdraw-footer">
+      <Button
+        text="confirm withdrawal"
+        width="full"
+        class="withdraw-btn"
+        disabled="true"
```
Comment (roma219) [inline src/views/Withdraw/WithdrawFinish.vue]: ?

File: src/views/Withdraw/WithdrawConfirmAmount.vue [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+      :centered="true"
+      :error="error"
+      :note-clickable="true"
+      title="withdrawal amount"
+      type="number"
+      @note-clicked="amount = withdrawAsset.tokens"
+    />
+    <div class="withdraw-footer">
+      <Button
+        :disabled="isAmountValid"
```
Comment (roma219) [inline src/views/Withdraw/WithdrawConfirmAmount.vue]: So it should be named `isAmountInvalid` i guess

**Review by roma219 — CHANGES_REQUESTED** [review]
1) This is the initial display:
image
For understanding how scrolling works with shadowers see `SortableTable` component
Shadower should be at the bottom aswell.
2) Button should be disabled while fetching for user is in progress, right now i can enter and click confirm on an non existing user
image
3) Why is button disabled here?
image
4) Title is not style correctly and padding is too big
image
5) Button is not placed correctly on mobile
image

**Review by roma219 — APPROVED** [review]

**Pull request merged**